src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.CompilationTracker.cs

```diff
@@ -180,6 +180,7 @@ SymbolKind.NetModule or
                         // And then this is the only place where we get a NoCompilationState and create a new instance.
                         // So there is no way to ever transition this to the DocumentsAreFinal=true state.
                         Contract.ThrowIfTrue(state.GeneratorInfo.DocumentsAreFinal);
+                        Contract.ThrowIfTrue(state.GeneratorInfo.Driver != null);
```

Comment:
had offline chat where i explained my reasoning. First, there are only two ways to get a NoCompilationState. Either because we were at the 'initial empty' state, or because we got into this codepath (we don't create a NoCompilationState anywhere else).

But this codepath checks if .Driver is null or not. But if you have the initial empty state, it is null. And as such, we never update the generator info. So the generator info always stays in the default state (it never get documents, and never changes the 'final' state, and never changse teh 'Driver' state). So we literally never mutate anything here.

Note: a larger cleanup is done in https://github.com/dotnet/roslyn/pull/71993.  But this PR is safe/reasonable to take in isolation.